doc: Clarify -blocksdir usage

sangaman opened the pull request:
This PR attempts to clarify and correct the `-blocksdir` argument description and default value. `-blocksdir` does not refer to the full path to the actual `blocks` directory, but rather the root/parent directory which contains the `blocks` directory. Accordingly, the default value is `<datadir>` and not `<datadir>/blocks` - this behavior of defaulting to the datadir can also be seen in init.cpp:

```cpp
    if (gArgs.IsArgSet("-blocksdir")) {
        path = fs::system_complete(gArgs.GetArg("-blocksdir", ""));
        if (!fs::is_directory(path)) {
            path = "";
            return path;
        }
    } else {
        path = GetDataDir(false);
    }
```

It also attempts to clarify that only the `.dat` files containing block data are impacted by `-blocksdir`, not the index files.

I believe this would close #12828.

promag commented on doc/man/bitcoin-qt.1:
```diff
@@ -39,7 +39,7 @@ Extra transactions to keep in memory for compact block reconstructions
 \fB\-blocksdir=\fR<dir>
 .IP
-Specify blocks directory (default: <datadir>/blocks)
+Specify blocks root directory for .dat files (default: <datadir>)
```
These are updated automatically, see https://github.com/bitcoin/bitcoin/commits/master/doc/man/bitcoind.1

promag reviewed (commented):
ACK, help message needs fixing.

> It also attempts to clarify that only the .dat files containing block data are impacted by -blocksdir, not the index files.

How does it clarify? It doesn't mention index files.

sangaman commented:
@promag Thank you! I wasn't sure how detailed/verbose to make the description. How about `Specify blocks root directory for .dat files, index .ldb files remain under datadir`?

sangaman committed:
doc: Clarify -blocksdir usage

This commit attempts to clarify and correct the `-blocksdir` argument
description and default value. `-blocksdir` does not refer to the full
path to the actual `blocks` directory, but rather the root/parent
directory which contains the `blocks` directory. Accordingly, the
default value is `<datadir>` and not `<datadir>/blocks`. It also
attempts to clarify that only the `.dat` files containing block data are
impacted by `-blocksdir`, not the index files.

hebasto commented on src/init.cpp:
```diff
@@ -333,7 +333,7 @@ void SetupServerArgs()
     gArgs.AddArg("-alertnotify=<cmd>", "Execute command when a relevant alert is received or we see a really long fork (%s in cmd is replaced by message)", false, OptionsCategory::OPTIONS);
     gArgs.AddArg("-assumevalid=<hex>", strprintf("If this block is in the chain assume that it and its ancestors are valid and potentially skip their script verification (0 to verify all, default: %s, testnet: %s)", defaultChainParams->GetConsensus().defaultAssumeValid.GetHex(), testnetChainParams->GetConsensus().defaultAssumeValid.GetHex()), false, OptionsCategory::OPTIONS);
-    gArgs.AddArg("-blocksdir=<dir>", "Specify blocks directory (default: <datadir>/blocks)", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-blocksdir=<dir>", "Specify blocks root directory for .dat files (default: <datadir>)", false, OptionsCategory::OPTIONS);
```
Could reference block data files as described in [doc/files.md](https://github.com/bitcoin/bitcoin/blob/master/doc/files.md): `blocks/*.dat` ?

sangaman replied:
Good idea but I'm not sure the best way to phrase that... `Specify root directory for blocks/*.dat files...`?

jonasschnelli replied:
I would reed this as the .dat files will be written to the "root directory" (but they get written into "<root>/blocks"). I think mentioning the "blocks" subdirectory (that will e created) makes sense.

hebasto replied:
```
"Specify directory to hold blocks subdirectory for *.dat files (default: <datadir>)"
```

sangaman commented:
Anything I should do to help get this merged? Let me know, I'd be happy to make any changes.

sangaman commented:
I went ahead and changed the description to `Specify root for blocks subdirectory to store .dat files` per the feedback I've gotten. I can't think of a clearer way to put it.

hebasto commented on src/httpserver.cpp:
```diff
@@ -278,7 +278,7 @@ static void http_reject_request_cb(struct evhttp_request* req, void*)
 }
 
-/** Event dispatcher thread */
+/** Event dispatcher thread */ x
```
Mistype?

sangaman replied:
Oops yes, sorry about that. Thanks!

sangaman commented:
@hebasto Thanks, that's a good way to put it I think.

hebasto commented:
utACK ccc27bdcd2d91fe2c023ad004019d5b970f21dbf

MarcoFalke commented on src/init.cpp:
```diff
@@ -333,7 +333,7 @@ void SetupServerArgs()
     gArgs.AddArg("-alertnotify=<cmd>", "Execute command when a relevant alert is received or we see a really long fork (%s in cmd is replaced by message)", false, OptionsCategory::OPTIONS);
     gArgs.AddArg("-assumevalid=<hex>", strprintf("If this block is in the chain assume that it and its ancestors are valid and potentially skip their script verification (0 to verify all, default: %s, testnet: %s)", defaultChainParams->GetConsensus().defaultAssumeValid.GetHex(), testnetChainParams->GetConsensus().defaultAssumeValid.GetHex()), false, OptionsCategory::OPTIONS);
-    gArgs.AddArg("-blocksdir=<dir>", "Specify blocks directory (default: <datadir>/blocks)", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-blocksdir=<dir>", "Specify directory to hold blocks subdirectory for *.dat files (default: <datadir>)", false, OptionsCategory::OPTIONS);
```
Not sure if `*.dat files` is worth to mention, maybe `block and undo data`?

MarcoFalke commented:
ACK

Sjors commented:
Concept ACK.

@MarcoFalke wrote:

> Not sure if `*.dat files` is worth to mention, maybe `block and undo data`?

I think `*.dat` is actually more clear and doesn't require understanding the inner workings.

This behavior is thoroughly confusing so I wouldn't worry about making the text a bit verbose.

I recommend adding an explanation that blocks/index stays in the datadir, because that's a source of mistakes when people think they can just move the entire `blocks` directory to some other disk.

So maybe:

```
-blocksdir=<blocksdir>
Specify directory to hold blocks subdirectory (default: <datadir>). The *.dat files will
be placed in <blocksdir>/blocks, but note that <datadir>/blocks/index remains in place for
performance reasons.
```

Longer term imo we should move `<datadir>/blocks/index` to `<datadir>/indexes/blocks`, and make `-blocksdir` a network specific setting.

sangaman commented:
I'd agree that being verbose is worth it if it makes things clearer, and I agree that `*.dat` files might be easier to grasp for users who are less familiar with the disk data of bitcoin and are unsure what `block and undo data` are - but ultimately I think either way is fine and a significant improvement over how its currently worded.

hebasto commented:
@Sjors 

> Longer term imo we should move `<datadir>/blocks/index` to `<datadir>/indexes/blocks`, and make `-blocksdir` a network specific setting.

#14409 plus network specific directories, right?

Sjors commented:
Ah, good to see `-blocksdir` was already made network specific. In that case we can drop the creation of a `/blocks` and `/testnet3/blocks` subdirectory (unless it already exists, for backwards compatibility).

sangaman commented:
Is there any consensus on whether the way it's currently worded is good? It certainly seems fine to me, but I would gladly change it if that would help move this along. I think anything would be a lot clearer than how that config option is currently described. Thanks.

MarcoFalke commented:
> Longer term imo we should move <datadir>/blocks/index to <datadir>/indexes/blocks, and make -blocksdir a network specific setting.

@Sjors Create an issue about that?
